api/api.go

```diff
@@ -0,0 +1,387 @@
+package api
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	types "github.com/thecodeteam/goscaleio/types/v1"
+)
+
+const (
+	HeaderKeyAccept                       = "Accept"
+	HeaderKeyContentType                  = "Content-Type"
+	HeaderValContentTypeJSON              = "application/json"
+	headerValContentTypeBinaryOctetStream = "binary/octet-stream"
+)
+
+var (
+	errNewClient = errors.New("missing endpoint")
+	errSysCerts  = errors.New("Unable to initialize cert pool from system")
+)
+
+// Client is an API client.
+type Client interface {
+
+	// Do sends an HTTP request to the API.
+	Do(
+		ctx context.Context,
+		method, path string,
+		body, resp interface{}) error
+
+	// DoWithHeaders sends an HTTP request to the API.
+	DoWithHeaders(
+		ctx context.Context,
+		method, path string,
+		headers map[string]string,
+		body, resp interface{}) error
+
+	// DoandGetREsponseBody sends an HTTP reqeust to the API and returns
+	// the raw response body
+	DoAndGetResponseBody(
+		ctx context.Context,
+		method, path string,
+		headers map[string]string,
+		body interface{}) (*http.Response, error)
+
+	// Get sends an HTTP request using the GET method to the API.
+	Get(
+		ctx context.Context,
+		path string,
+		headers map[string]string,
+		resp interface{}) error
+
+	// Post sends an HTTP request using the POST method to the API.
+	Post(
+		ctx context.Context,
+		path string,
+		headers map[string]string,
+		body, resp interface{}) error
+
+	// Put sends an HTTP request using the PUT method to the API.
+	Put(
+		ctx context.Context,
+		path string,
+		headers map[string]string,
+		body, resp interface{}) error
+
+	// Delete sends an HTTP request using the DELETE method to the API.
+	Delete(
+		ctx context.Context,
+		path string,
+		headers map[string]string,
+		resp interface{}) error
+
+	// SetToken sets the Auth token for the HTTP client
+	SetToken(token string)
+
+	// GetToken gets the Auth token for the HTTP client
+	GetToken() string
+
+	// ParseJSONError parses the JSON in r into an error object
+	ParseJSONError(r *http.Response) error
+}
+
+type client struct {
+	http     *http.Client
+	host     string
+	token    string
+	showHTTP bool
+}
+
+// ClientOptions are options for the API client.
+type ClientOptions struct {
+	// Insecure is a flag that indicates whether or not to supress SSL errors.
+	Insecure bool
+
+	// UseCerts is a flag that indicates whether system certs should be loaded
+	UseCerts bool
+
+	// Timeout specifies a time limit for requests made by this client.
+	Timeout time.Duration
+
+	// ShowHTTP is a flag that indicates whether or not HTTP requests and
+	// responses should be logged to stdout
+	ShowHTTP bool
+}
+
+// New returns a new API client.
+func New(
+	ctx context.Context,
+	host string,
+	opts *ClientOptions) (Client, error) {
```

Comment:
Hi @codenrhoden,

Consider passing `opts *ClientOptions` by value instead?

```go
func New(
 	ctx context.Context,
 	host string,
 	opts ClientOptions) (Client, error) {
```

You would no longer need to check if `opts` is `nil`, and the best part is that the configuration properties in `opts` all have default values equivalent to if `opts` was `nil` in the first place :)